### PR TITLE
ONNX Export Support for CrossEntropyLoss

### PR DESCRIPTION
**Pull request by KsenijaS**
Add ONNX export support for torch.nn.CrossEntropyLoss.

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -2617,6 +2617,65 @@ def forward(self, *tensor_list):
         x = torch.randn(3, 4)
         self.run_test(EinsumModelTranspose(), input=(x,))
 
+    @skipIfUnsupportedMinOpsetVersion(12)
```
Comment (lara-hdr) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: also temporarly add something like : @unittest.skip("Enable this once ORT version is updated")

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -2617,6 +2617,63 @@ def forward(self, *tensor_list):
         x = torch.randn(3, 4)
         self.run_test(EinsumModelTranspose(), input=(x,))
 
+    @unittest.skip("Enable this once ORT version is updated")
```
Comment (lara-hdr) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: could you add 1 or 2 tests in test/onnx/test_operators.py as well, to test the graph schema?

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 3]
```diff
@@ -815,6 +815,32 @@ void removeMaxPoolUnusedOutput(Block* b) {
   }
 }
 
+static void fuseLogSoftmaxNllLoss(Block* b) {
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: Looks good to me. Maybe add some comments on what the pass does.

**Comment by lara-hdr** [comment]
OnnxBackendSimpleModelTest.test_gradient_of_add_and_mul_cpu is failing on the CI, could you run it locally and see if you can repro the error?

**Review by lara-hdr — APPROVED** [review]
LGTM.Thanks!

**Comment by KsenijaS** [comment]
@houseroad can you please review?

**Review by houseroad — APPROVED** [review]
LG

**Review by neginraoof — APPROVED** [review]
LGTM.

**Comment by KsenijaS** [comment]
@houseroad can you please merge this PR if there are no more comments

**Comment by ezyang** [comment]
It looks like celu cuda needed to be blacklisted too

```
21:01:33 =================================== FAILURES ===================================
21:01:33 ___________________ OnnxBackendNodeModelTest.test_celu_cuda ____________________
21:01:33 
21:01:33 test_self = <caffe2.python.onnx.tests.onnx_backend_test.OnnxBackendNodeModelTest testMethod=test_celu_cuda>
21:01:33 device = 'CUDA'
21:01:33 
21:01:33     def run(test_self, device):  # type: (Any, Text) -> None
21:01:33         if model_test.model_dir is None:
21:01:33             model_dir = self.prepare_model_data(model_test)
21:01:33         else:
21:01:33             model_dir = model_test.model_dir
21:01:33         model_pb_path = os.path.join(model_dir, 'model.onnx')
21:01:33         model = onnx.load(model_pb_path)
21:01:33         model_marker[0] = model
21:01:33         if hasattr(self.backend, 'is_compatible') \
21:01:33            and callable(self.backend.is_compatible) \
21:01:33            and not self.backend.is_compatible(model):
21:01:33             raise unittest.SkipTest('Not compatible with backend')
21:01:33 >       prepared_model = self.backend.prepare(model, device)
21:01:33 
21:01:33 ../.local/lib/python3.6/site-packages/onnx/backend/test/runner/__init__.py:278: 
21:01:33 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
21:01:33 ../.local/lib/python3.6/site-packages/caffe2/python/onnx/backend.py:713: in prepare
21:01:33     init_net, predict_net = cls._onnx_model_to_caffe2_net(model, device, opset_version, False)
21:01:33 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
21:01:33 
21:01:33 cls = <class 'caffe2.python.onnx.backend.Caffe2Backend'>
21:01:33 onnx_model = ir_version: 7
21:01:33 producer_name: "backend-test"
21:01:33 graph {
21:01:33   node {
21:01:33     input: "X"
21:01:33     output: "Y"
21:01:33     op_type: "Celu"
21:01:33     at...  dim {
21:01:33             dim_value: 1
21:01:33           }
21:01:33         }
21:01:33       }
21:01:33     }
21:01:33   }
21:01:33 }
21:01:33 opset_import {
21:01:33   domain: ""
21:01:33   version: 12
21:01:33 }
21:01:33 
21:01:33 device = 'CUDA', opset_version = 12, include_initializers = False
21:01:33 
21:01:33     @classmethod
21:01:33     def _onnx_model_to_caffe2_net(cls, onnx_model, device, opset_version, include_initializers):
21:01:33         device_option = get_device_option(Device(device))
21:01:33     
21:01:33         onnx_model = onnx.utils.polish_model(onnx_model)
21:01:33         init_model = cls.optimize_onnx(onnx_model, init=True)
21:01:33         pred_model = cls.optimize_onnx(onnx_model, predict=True)
21:01:33     
21:01:33         init_net = caffe2_pb2.NetDef()
21:01:33         pred_net = caffe2_pb2.NetDef()
21:01:33     
21:01:33         init_net.name = onnx_model.graph.name + '_init'
21:01:33         pred_net.name = onnx_model.graph.name + '_predict'
21:01:33     
21:01:33         if include_initializers:
21:01:33             init_net.op.extend(cls._create_tensor_filling_op(tp) for tp in onnx_model.graph.initializer)
21:01:33     
21:01:33         cls._dummy_name.reset(cls._all_names_in_graph(init_model.graph) | cls._all_names_in_graph(pred_model.graph))
21:01:33     
21:01:33         errors = []
21:01:33         for net, model in ( (init_net, init_model), (pred_net, pred_model) ):
21:01:33             net.device_option.CopyFrom(device_option)
21:01:33             for node in model.graph.node:
21:01:33                 try:
21:01:33                     c2ops = cls._onnx_node_to_caffe2_op(
21:01:33                         init_model, pred_model, node, opset_version)
21:01:33                 except Exception as e:
21:01:33                     msg = 'Error while processing node: {}. Exception: {}'.format(node, e)
21:01:33                     errors.append(msg)
21:01:33                     print('ONNX FATAL:', msg, file=sys.stderr)
21:01:33                     continue
21:01:33                 init_net.op.extend(c2ops.init_ops)
21:01:33                 net.op.extend(c2ops.ops)
21:01:33                 net.external_input.extend(c2ops.interface_blobs)
21:01:33             net.external_output.extend(
21:01:33                 value_info.name for value_info in model.graph.output)
21:01:33             net.external_input.extend(
21:01:33                 value_info.name for value_info in model.graph.input)
21:01:33     
21:01:33         if len(errors) > 0:
21:01:33             raise RuntimeError(
21:01:33                 "ONNX conversion failed, encountered {} errors:\n\n{}".format(
21:01:33 >                   len(errors), "\n\n".join(errors)))
21:01:33 E           RuntimeError: ONNX conversion failed, encountered 1 errors:
21:01:33 E           
21:01:33 E           Error while processing node: input: "X"
21:01:33 E           output: "Y"
21:01:33 E           op_type: "Celu"
21:01:33 E           attribute {
21:01:33 E             name: "alpha"
21:01:33 E             f: 2.0
21:01:33 E             type: FLOAT
21:01:33 E           }
21:01:33 E           . Exception: Don't know how to translate op Celu
21:01:33 
```

**Comment by houseroad** [comment]
@ezyang which ci runs this cuda test?

**Comment by ezyang** [comment]
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py3.6-devtoolset7-rocmrpm-centos7-test/12913//console

**Pull request closed without merging**